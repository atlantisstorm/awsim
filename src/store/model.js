import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const handleFetchData = ({ state, accessKeyId, secretAccessKey, region }) => {
  let instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D
  if (process.env.NODE_ENV === "development") {
    instances = extractInstances(testInstances);
  } else {
    AWS.config.update({
      "accessKeyId": accessKeyId,
      "secretAccessKey": secretAccessKey,
      "region": region
    });

    const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});
    const params = {
      DryRun: false
    };

    ec2.describeInstances(params, function(err, data) {
      if(err) {
        console.log("Error", err.stack);
      } else {
        instances = extractInstances(data);
      }
    });
  }

  return { ...state, loading: false, instances };
}

export const handleFilterInstances = ({ state }) => {
  const filteredInstances = filterInstances(state.instances, state.filters);
  return { ...state, filteredInstances };
}

export const handleFilterUpdate = ({ state, filterName, filterValue }) => {
  const filters = state.filters;
  const updatedFilter = { ...filters[filterName], currentValue: filterValue };
  const updatedFilters = { ...filters, [filterName]: updatedFilter };
  const filteredInstances = filterInstances(state.instances, updatedFilters);

  return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };    
}

export const handleModalSetDisplay = ({ state, display }) => {
  const newModalSettings = {
    ...state.modalSettings,
    display: display
  };

  return { ...state, modalSettings: newModalSettings };
}

export const handleModalShowInstance = ({ state, instanceId, pageX, pageY}) => {
  const newModalSettings = {
    ...state.modalSettings,
    pageX,
    pageY,
    display: true,    
    instance: getInstanceRawData(state.instances, instanceId)
  };

  return { ...state, selectedInstance: instanceId, modalSettings: newModalSettings };
}

export const handleRefreshData = ({ state }) => {
  return { ...state, loading: true };
}

export const handleSortOptionsSet = ({ state, field, order }) => {
  return { ...state, sortOptions: setSort(field, order) };
}

export const Model = {
  handleFetchData,
  handleFilterInstances,
  handleFilterUpdate,
  handleModalSetDisplay,
  handleModalShowInstance,
  handleRefreshData,
  handleSortOptionsSet
};

export default Model;