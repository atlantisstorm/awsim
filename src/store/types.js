// Emmm... action types.
export const FETCH_DATA = "FETCH_DATA";
export const FILTER_INSTANCES = "FILTER_INSTANCES";
export const FILTER_UPDATE = "FILTER_UPDATE";
export const MODAL_SET_DISPLAY = "MODAL_SET_DISPLAY";
export const MODAL_SHOW_INSTANCE = "MODAL_SHOW_INSTANCE";
export const REFRESH_DATA = "REFRESH_DATA";
export const SORT_OPTIONS_SET = "SORT_OPTIONS_SET";