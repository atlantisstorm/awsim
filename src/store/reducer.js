import { useReducer } from 'react';
import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
import {
  FETCH_DATA,
  FILTER_INSTANCES,
  FILTER_UPDATE,
  MODAL_SET_DISPLAY,
  MODAL_SHOW_INSTANCE,
  REFRESH_DATA,
  SORT_OPTIONS_SET
} from './types';

const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const reducer = (state, action) => {
  let payload;
  let instances;
  let filters;
  let newModalSettings;

  switch (action.type) {
    case FETCH_DATA: {
      payload = action.payload;
      instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D

      if (process.env.NODE_ENV === "development") {
        instances = extractInstances(testInstances);
      } else {
        AWS.config.update({
          "accessKeyId": payload.accessKeyId,
          "secretAccessKey": payload.secretAccessKey,
          "region": payload.region
        });

        const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});

        const params = {
          DryRun: false
        };

        ec2.describeInstances(params, function(err, data) {
          if(err) {
            console.log("Error", err.stack);
          } else {
            instances = extractInstances(data);
          }
        });
      }

      return { ...state, loading: false, instances };
    }

    case FILTER_INSTANCES: {
      const filteredInstances = filterInstances(state.instances, state.filters);
      return { ...state, filteredInstances };
    }

    case FILTER_UPDATE: {
      payload = action.payload;
      filters = state.filters;

      const filterName = payload.name;
      const filterValue = payload.value;
      const updatedFilter = { ...filters[filterName], currentValue: filterValue };
      const updatedFilters = { ...filters, [filterName]: updatedFilter };

      const filteredInstances = filterInstances(state.instances, updatedFilters);
      return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };
    }

    case MODAL_SET_DISPLAY: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: payload.display
      };
      return { ...state, modalSettings: newModalSettings };
    }

    case MODAL_SHOW_INSTANCE: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: true,
        pageX: payload.pageX, 
        pageY: payload.pageY,
        instance: getInstanceRawData(state.instances, payload.instanceId)
      };
      return { ...state, selectedInstance: payload.instanceId, modalSettings: newModalSettings };
    }

    case REFRESH_DATA: {
      // This will trigger FETCH_DATA
      return { ...state, loading: true };
    }

    case SORT_OPTIONS_SET: {
      // This will trigger FETCH_DATA
      payload = action.payload;
      return { ...state, sortOptions: setSort(payload.field, payload.order) };
    }

    default: {
      return state;
    }
  }
};

export const initialiseReducer = (props) => {
  const iState = (props && props.initialState) ? props.initialState : defaultState;
  const [state, dispatch] = useReducer(reducer, iState);
  return { state, dispatch };
}