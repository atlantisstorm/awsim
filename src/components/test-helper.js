import React from 'react';
//import { render } from "@testing-library/react";
import Context from "../store/context";
import { initialiseReducer } from "../store/reducer";

export const RenderWithProvider = (props) => {
  const { children, testState, testDispatch }  = props;
  // Note: we pass any supplied testState to initialiseReducer rather than switching it like
  // we do for dispatch because of scoping.  Basically each rerender after a click/select
  // update event would mean we refresh again with the original data.
  const { state, dispatch: iDispatch } = initialiseReducer({ initialState: testState });
  const dispatch = testDispatch ? testDispatch : iDispatch;
  return (
    <Context.Provider value={{ state, dispatch }}>
      { children }
    </Context.Provider> 
  )
}