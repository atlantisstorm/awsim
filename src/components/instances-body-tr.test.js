import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import InstancesBodyTr from './instances-body-tr';
import extractInstances from '../fixtures/extracted-instances';
import { defaultState } from '../store/reducer';
import { MODAL_SHOW_INSTANCE } from '../store/types';
import { RenderWithProvider } from './test-helper';

describe('InstancesBodyTr component', () => {
  const testState = defaultState

  const sortField = "Name";
  const extractedInstance = extractInstances[0];

  it('should match snapshot', () => {
    const { asFragment } = render(
      <RenderWithProvider testState={testState}>
        <table>
          <tbody>
            <InstancesBodyTr sortField={sortField} instance={extractedInstance} />
          </tbody>
        </table>
      </RenderWithProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call dispatch with expected parameters when tr row is clicked', () => {
    const dispatch = jest.fn();
    const { getByTestId } = render(
      <RenderWithProvider testDispatch={dispatch} testState={testState}>
        <table>
          <tbody>
            <InstancesBodyTr sortField={sortField} instance={extractedInstance} />
          </tbody>
        </table>
      </RenderWithProvider>
    );

    const showInstanceDetails = getByTestId('show-instance-details');
    fireEvent.click(showInstanceDetails);
    const payload = {
      "instanceId": "",
      "pageX": 0,
      "pageY": 0
    };
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: MODAL_SHOW_INSTANCE, payload });
  });
});