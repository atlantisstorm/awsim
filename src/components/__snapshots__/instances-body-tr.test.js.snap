// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstancesBodyTr component should match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class=""
        data-instance-id="i-07ae1e2a36d2b88bf"
        data-testid="show-instance-details"
      >
        <td
          class="sorted"
        >
          ftp-dev.awsbox.es
        </td>
        <td
          class=""
        >
          dev
        </td>
        <td
          class=""
        >
          running
        </td>
        <td
          class="launch-ssh js-launch-ssh"
          data-ip-address="54.42.117.131"
          data-target-platform=""
          title="Launch remote connection"
        >
          54.42.117.131
        </td>
        <td
          class="launch-ssh js-launch-ssh"
          data-ip-address="192.168.34.117"
          data-target-platform=""
          title="Launch remote connection"
        >
          192.168.34.117
        </td>
        <td
          class=""
        />
        <td
          class=""
        />
        <td
          class=""
        >
          i-07ae1e2a36d2b88bf
        </td>
        <td
          class=""
        >
          ami-5442e32fd1ababf88
        </td>
        <td
          class=""
        >
          t2.micro
        </td>
        <td
          class=""
        >
          2019-08-14 15:08:39
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
